Group-Object - Use Case-Sensitive Hashtable for -CaseSensitive -AsHashtable

vexx32 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Prior to this change, `Group-Object -AsHashtable -CaseSensitive` would give a key duplication error when given entries that only differ by casing. This was due to always using a case-insensitive hashtable, despite the request for `-CaseSensitive` behaviour.

This change allows `Group-Object -CaseSensitive -AsHashtable` to create and return a case-sensitive hashtable to enable this use case.

## PR Context

Resolves #10441

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

vexx32 committed:
:sparkles: :white_check_mark: Add case sensitivity

- Group-Object -AsHashtable -CaseSensitive returns a case-sensitive hash
instead of erroring out when given keys differing only by case.

- Added tests

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Group-Object.cs:
```diff
@@ -518,7 +518,16 @@ protected override void EndProcessing()
+                    }
+                    else
+                    {
+                        hashtable = CollectionsUtil.CreateCaseInsensitiveHashtable();
```
I think we can simplify the code and this helps us to add Culture parameter later #9348:
```c#
var comparer = CaseSensitive.IsPresent ? StringComparer.CurrentCulture : StringComparer.CurrentCultureIgnoreCase;
hashtable = Hashtable(comparer);

vexx32 replied:
Can do! I'm not sure if these methods use currentculture or ordinal though. I'll check :)

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1:
```diff
@@ -7,9 +7,9 @@ Describe "Group-Object DRT Unit Tests" -Tags "CI" {
-        ,$results | Should -BeOfType "System.Array"
+        $results.Name | Should -Be aA, AA
+        $results.Group | Should -Be aA, aA, AA, AA
+        , $results | Should -BeOfType "System.Array"
```
Please move style fixes in another PR. Below too.

vexx32 replied:
VS code's autoformat strikes again. It is in a separate commit at least, but if you'd prefer to do a whole separate PR I can manage. :slightly_smiling_face:

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1:
```diff
@@ -173,6 +173,23 @@ Describe "Check 'Culture' parameter in order object cmdlets (Group-Object, Sort-
+            $capitonyms | Group-Object Capitonym -AsHashTable -CaseSensitive
+        }
+
+        $script | Should -Not -Throw
```
I'd prefer positive checks.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1:
```diff
@@ -173,6 +173,23 @@ Describe "Check 'Culture' parameter in order object cmdlets (Group-Object, Sort-
+                }
+            )
+
+            $capitonyms | Group-Object Capitonym -AsHashTable -CaseSensitive
```
```suggestion
            $capitonyms | Group-Object -Property Capitonym -AsHashTable -CaseSensitive
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Group-Object.cs:
```diff
@@ -518,7 +518,16 @@ protected override void EndProcessing()
             {
                 if (AsHashTable)
```
For consistency:
```suggestion
                 if (AsHashTable.IsPresent)
```

vexx32 committed:
:recycle: Refactor GroupObjectCommand

- Consolidate hashtable creation

- Better consistency of handling switches in conditionals

vexx32 committed:
:white_check_mark: Update tests

vexx32 commented:
@iSazonov I think that should address all your concerns. I'll look at doing a style pass over the tests and possible the cmdlet after this PR is completed. 😊

iSazonov commented:
@vexx32 I don't know should we document this in docs repo?

vexx32 commented:
🤔 I suppose it _should_ be documented since we're changing established behaviour. I'll open an issue.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1:
```diff
@@ -173,6 +173,20 @@ Describe "Check 'Culture' parameter in order object cmdlets (Group-Object, Sort-
+            }
+        )
+
+        [hashtable] $Result = $capitonyms | Group-Object -Property Capitonym -AsHashTable -CaseSensitive
```
If this gets regressed in the future, I think this will give a better test error message

```suggestion
        $Result = $capitonyms | Group-Object -Property Capitonym -AsHashTable -CaseSensitive
        $Result | Should -BeOfType [HashTable]
```

vexx32 replied:
Good catch! 🙂

vexx32 committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/Group-Object.Tests.ps1

Co-Authored-By: Steve Lee <slee@microsoft.com>

vexx32 commented:
@SteveL-MSFT 🤔 do you know if the bot ignores non-required checks? Pretty sure it's waiting on the CodeFactor issue (which is just "complex method" on EndProcessing() here.

I'm happy to refactor it out into helper methods if you feel that's necessary. 🙂
